### PR TITLE
bump spack version

### PR DESCRIPTION
**Pull request by samuelpmish**
Some of the [build_issues](https://github.com/spack/spack/issues/12238) we have on Ubuntu/WSL are related to using an older version of spack.

Other potential fixes:
-target the master branch and do to not specify a specific commit
-add spack as a submodule, rather than use uberenv to download it

### CODE REVIEW

**Comment by white238** [comment]
Bumping Spack versions is a very risky thing.  There are a lot of moving parts inside of Spack.  It has been my experience that anytime you upgrade Spack, it breaks your TPLs because someone has submitted a fix for a common package for their system that doesn't work for ours.

This is why we lock down the Spack version through Uberenv to a very specific and vetted SHA.  

Also anytime we bump the version, whoever does this needs to make sure it builds on both rzgenie and quartz for our set of compiler suites.  This is made easier with the scripts/llnl/build_tpls.py script.  The new host-configs should be updated inside of the PR that increases the version.

**Comment by adrienbernede** [comment]
LGTM

**Comment by adrienbernede** [comment]
(Let’s see if bumping spack version works on quartz ...)

**Comment by adrienbernede** [comment]
Per @white238 comment, I don’t know where this is going, but it passes CI on quartz. (Builds OK, same failures in tests).

**Comment by adrienbernede** [comment]
To complete @white238 and share some recent findings:
Uberenv patches spack default config quite brutally, replacing it with the one the project provides.
The consequence is that some default values may not be set because the patch config may be outdated.

**Comment by samuelpmish** [comment]
Can someone trigger the CI tools to check that this builds and runs without error on genie and quartz?

**Comment by adrienbernede** [comment]
@samuelpmish 
For quartz CI, you can trigger it yourself with a "LGTM" sing-off comment. I’ll let you do it :)

**Comment by samuelpmish** [comment]
@adrienbernede I only have access to my personal (i.e. non-LLNL) github account, is it still enabled to perform the sign-off?

**Comment by samuelpmish** [comment]
LGTM

**Comment by jamiebramwell** [comment]
yeah, it needs to be your blessed LLNL account. I can kick it off.

**Comment by jamiebramwell** [comment]
LGTM

**Comment by adrienbernede** [comment]
I didn’t noticed...

**Comment by jamiebramwell** [comment]
Looks like it failed with clang.

**Comment by adrienbernede** [comment]
I’m restarting the failing job...

**Comment by adrienbernede** [comment]
That’s what I thought... This failure is not reproducible, and I don’t know its origin.

File: .vscode/launch.json [hunk 1]
```diff
@@ -1,20 +1,111 @@
...
+                }
+            ]
+        },
+        {
+            "name": "(gdb) mpi Dynamic Test",
+            "type": "cppdbg",
+            "request": "launch",
+            "program": "${workspaceFolder}/uberenv_libs/gcc-7.4.0/mpich-3.3.2/bin/mpiexec",
+            "args": [
+                "-n", "4", "/home/sam/serac/build/tests/serac_dynamic_solver",
```
Comment (white238) [inline .vscode/launch.json:29]: These hardcoded paths are concerning.

File: host-configs/quartz-toss_3_x86_64_ib-clang@4.0.0.cmake [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+set(ENABLE_MPI "ON" CACHE PATH "")
+
+set(MPI_C_COMPILER "/usr/tce/packages/mvapich2/mvapich2-2.2-clang-4.0.0/bin/mpicc" CACHE PATH "")
+
+set(MPI_CXX_COMPILER "/usr/tce/packages/mvapich2/mvapich2-2.2-clang-4.0.0/bin/mpicxx" CACHE PATH "")
+
+#---------------------------------------
+# Library Dependencies
+#---------------------------------------
+set(TPL_ROOT "/usr/WS2/smithdev/libs/toss_3_x86_64_ib/2020_03_25_23_14_51/clang-4.0.0" CACHE PATH "")
```
Comment (white238) [inline host-configs/quartz-toss_3_x86_64_ib-clang@4.0.0.cmake:36]: Are you rebuilding these with the new spack?

**Comment by white238** [comment]
Jamie had an idea that this might be diffing against an out of date master due to the Github outage earlier today.  Can you make a small change and commit it to see if it updates?

File: .vscode/launch.json (hunk 1 above)
Comment (samuelpmish) [reply .vscode/launch.json:29]: sorry, these files will just be removed

**Comment by samanseifi** [comment]
So I build `serac` with dependencies and everything according to the fix in issue #44. Tests are failed at the end 6 out of 8. Is this the same issue here?
 
```
Test project /home/saman/serac/build-saman-Latitude-E-linux-ubuntu18.04-broadwell-gcc@7.5.0-debug
    Start 1: blt_gtest_smoke
1/8 Test #1: blt_gtest_smoke ..................   Passed    0.00 sec
    Start 2: blt_mpi_smoke
2/8 Test #2: blt_mpi_smoke ....................***Failed    0.05 sec
    Start 3: serac_dynamic_solver
3/8 Test #3: serac_dynamic_solver .............***Failed    0.02 sec
    Start 4: serac_linelastic_solver
4/8 Test #4: serac_linelastic_solver ..........***Failed    0.02 sec
    Start 5: serac_quasistatic_solver
5/8 Test #5: serac_quasistatic_solver .........***Failed    0.01 sec
    Start 6: serac_thermal_solver
6/8 Test #6: serac_thermal_solver .............***Failed    0.02 sec
    Start 7: serac_dtor
7/8 Test #7: serac_dtor .......................***Failed    0.01 sec
    Start 8: stdfunction_coefficient
8/8 Test #8: stdfunction_coefficient ..........   Passed    0.11 sec

25% tests passed, 6 tests failed out of 8

Total Test time (real) =   0.26 sec

The following tests FAILED:
	  2 - blt_mpi_smoke (Failed)
	  3 - serac_dynamic_solver (Failed)
	  4 - serac_linelastic_solver (Failed)
	  5 - serac_quasistatic_solver (Failed)
	  6 - serac_thermal_solver (Failed)
	  7 - serac_dtor (Failed)
Errors while running CTest
Makefile:129: recipe for target 'test' failed
make: *** [test] Error 8
```

**Comment by jamiebramwell** [comment]
@samanseifi My best guess is it's a problem with your MPI since only the MPI tests fail. What is the output of ` ctest -R blt_mpi_smoke --verbose` ?

**Comment by samanseifi** [comment]
Yeah I think so. That's what I get:

```
UpdateCTestConfiguration  from :/home/saman/serac/build-saman-Latitude-E-linux-ubuntu18.04-broadwell-gcc@7.5.0-debug/DartConfiguration.tcl
Parse Config file:/home/saman/serac/build-saman-Latitude-E-linux-ubuntu18.04-broadwell-gcc@7.5.0-debug/DartConfiguration.tcl
UpdateCTestConfiguration  from :/home/saman/serac/build-saman-Latitude-E-linux-ubuntu18.04-broadwell-gcc@7.5.0-debug/DartConfiguration.tcl
Parse Config file:/home/saman/serac/build-saman-Latitude-E-linux-ubuntu18.04-broadwell-gcc@7.5.0-debug/DartConfiguration.tcl
Test project /home/saman/serac/build-saman-Latitude-E-linux-ubuntu18.04-broadwell-gcc@7.5.0-debug
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 2
    Start 2: blt_mpi_smoke

2: Test command: /home/saman/serac/uberenv_libs/gcc-7.5.0/openmpi-3.1.5/bin/mpiexec "-n" "4" "/home/saman/serac/build-saman-Latitude-E-linux-ubuntu18.04-broadwell-gcc@7.5.0-debug/tests/blt_mpi_smoke"
2: Test timeout computed to be: 1500
2: --------------------------------------------------------------------------
2: There are not enough slots available in the system to satisfy the 4
2: slots that were requested by the application:
2: 
2:   /home/saman/serac/build-saman-Latitude-E-linux-ubuntu18.04-broadwell-gcc@7.5.0-debug/tests/blt_mpi_smoke
2: 
2: Either request fewer slots for your application, or make more slots
2: available for use.
2: 
2: A "slot" is the Open MPI term for an allocatable unit where we can
2: launch a process.  The number of slots available are defined by the
2: environment in which Open MPI processes are run:
2: 
2:   1. Hostfile, via "slots=N" clauses (N defaults to number of
2:      processor cores if not provided)
2:   2. The --host command line parameter, via a ":N" suffix on the
2:      hostname (N defaults to 1 if not provided)
2:   3. Resource manager (e.g., SLURM, PBS/Torque, LSF, etc.)
2:   4. If none of a hostfile, the --host command line parameter, or an
2:      RM is present, Open MPI defaults to the number of processor cores
2: 
2: In all the above cases, if you want Open MPI to default to the number
2: of hardware threads instead of the number of processor cores, use the
2: --use-hwthread-cpus option.
2: 
2: Alternatively, you can use the --oversubscribe option to ignore the
2: number of available slots when deciding the number of processes to
2: launch.
2: --------------------------------------------------------------------------
1/1 Test #2: blt_mpi_smoke ....................***Failed    0.04 sec

0% tests passed, 1 tests failed out of 1

Total Test time (real) =   0.04 sec

The following tests FAILED:
	  2 - blt_mpi_smoke (Failed)
Errors while running CTest
```

**Comment by jamiebramwell** [comment]
Yep, that looks like the problem. Most of our MPI tests use 4 processes. You can either modify your MPI install as noted in the error message, or locally modify your `tests/CMakeLists.txt` and reduce the `NUM_MPI_TASKS` variable.

**Comment by jamiebramwell** [comment]
> 
> 
> Yep, that looks like the problem. Most of our MPI tests use 4 processes. You can either modify your MPI install as noted in the error message, or locally modify your `tests/CMakeLists.txt` and reduce the `NUM_MPI_TASKS` variable.

@white238 @adrienbernede Please correct me if there is a more elegant solution.

**Comment by samanseifi** [comment]
So I changed the `NUM_MPI_TASKS` all to 1, now all passes but test 2 still fails. `ctest -R blt_mpi_smoke --verbose` produces same error. Apparently `blt_mpi_smoke` is still set to 4 processes. How can I fix this or even fix the MPI for 4 processes?

Where in the `serac` building workflow the MPI fix should be added?

**Comment by jamiebramwell** [comment]
> 
> 
> Where in the `serac` building workflow the MPI fix should be added?

I'm not really sure. I'll open a new ticket for this because it would be nice for it to be a user configurable parameter. @white238 would probably be the one who knows best.

**Comment by adrienbernede** [comment]
Changing `NUM_MPI_TASKS` in Serac tests do not affect the MPI smoke test because the latter is provided by BLT. 
It is possible to change `NUM_MPI_TASKS` in `serac/cmake/blt/tests/smoke/CMakeLists.txt` in the MPI Test, although that does not make much sense.

It seems to me that the best here is to disable MPI since apparently the goal is to run the code sequentially.

**Comment by white238** [comment]
@samanseifi If I am reading what you said correctly, either you have an install of MPI that is not configured correctly (ie. it cant run anything with parallelism, which is the whole point of MPI) or you want a serial version of Serac.

The latter is something we as a project need to decide whether it is worth supporting considering we normally run on LLNL super computers and laptops capable of at least some (4+) MPI tasks.

**Comment by samanseifi** [comment]
> @samanseifi If I am reading what you said correctly, either you have an install of MPI that is not configured correctly (ie. it cant run anything with parallelism, which is the whole point of MPI) or you want a serial version of Serac.
> 
> The latter is something we as a project need to decide whether it is worth supporting considering we normally run on LLNL super computers and laptops capable of at least some (4+) MPI tasks.

I followed the workflow [here](https://serac.readthedocs.io/en/latest/) and 
an MPI is installed local to `serac` installation. My `cmake` file is pointing at it:
```
set(ENABLE_MPI "ON" CACHE PATH "")

set(MPI_C_COMPILER "/home/saman/serac/uberenv_libs/gcc-7.5.0/openmpi-3.1.5/bin/mpicc" CACHE PATH "")

set(MPI_CXX_COMPILER "/home/saman/serac/uberenv_libs/gcc-7.5.0/openmpi-3.1.5/bin/mpic++" CACHE PATH "")

set(MPIEXEC_EXECUTABLE "/home/saman/serac/uberenv_libs/gcc-7.5.0/openmpi-3.1.5/bin/mpiexec" CACHE PATH "")
```
So yes the MPI is not configured correctly. But how to fix it within the `serac` installation workflow? What am  I missing here? A lot probably. :)

**Comment by white238** [comment]
@samanseifi If you have a working MPI install, you can add/alter your MPI install to scripts/uberenv/spack_configs/linux-ubuntu18.04-broadwell/packages.yaml.  There are examples of this in the other packages.yaml, you can follow.  If you have one installed by Spack you can use the "--upstream" option to uberenv.py.

This spack_config works for other people using the WSL for Ubuntu but not a lot so it may need to be tweaks a bit.  Just to be clear, do you have more than one core on your machine?

Also this PR is not really the place to debug this.  Please re-open your previous issue or create a new one so this PR can move forward.

**Comment by jamiebramwell** [comment]
@samanseifi  [Reopened here](https://github.com/LLNL/serac/issues/44).

File: tests/serac_dtor.cpp [hunk 3]
```diff
@@ -16,6 +14,12 @@ T do_nothing(T foo)
   return foo;
 }
 
+inline bool file_exists(const char *path)
+{
+  struct stat buffer;
+  return (stat(path, &buffer) == 0);
+}
+
```
Comment (samuelpmish) [inline tests/serac_dtor.cpp:24]: this doesn't seem to be used

File: tests/serac_dtor.cpp [hunk 4]
```diff
@@ -24,6 +28,22 @@ int main(int argc, char **argv)
...
+      args.PrintUsage(std::cout);
+    }
+    MPI_Finalize();
+    return 1;
+  }
+  if (myid == 0) {
+    args.PrintOptions(std::cout);
+  }
+
   std::ifstream imesh("../../data/beam-hex.mesh");
```
Comment (samuelpmish) [inline tests/serac_dtor.cpp:49]: `std::ifstream imesh(mesh_file)`

`std::ifstream` already has a way to check if the file exists, maybe we can move away from the platform-specific implementation of `file_exists`

**Comment by white238** [comment]
This passes all TPL builds, builds with the new host-configs, and passes all tests on quartz and rzgenie with our set of compilers.  This should be good to merge now.

**Comment by white238** [comment]
LGTM

**Pull request merged**